csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -0,0 +1,506 @@
+            }
+            catch (Exception)
+            {
+                DisposeHelper(true);
+                throw;
+            }
+        }
+
+        private string GetOutputName(ulong index, bool training)
+        {
+            var allocator = OrtAllocator.DefaultInstance;
+            IntPtr nameHandle;
+            string str = null;
+            if (training)
+            { 
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetTrainingModelOutputName(
+                                           _nativeHandle,
+                                           (UIntPtr)index,
+                                           allocator.Pointer,
+                                           out nameHandle));
+            } 
+            else
+            { 
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetEvalModelOutputName(
+                                           _nativeHandle,
+                                           (UIntPtr)index,
+                                           allocator.Pointer,
+                                           out nameHandle));
+            }
+
+            using (var ortAllocation = new OrtMemoryAllocation(allocator, nameHandle, 0))
+            {
+                str = NativeOnnxValueHelper.StringFromNativeUtf8(nameHandle);
+            }
+
+            return str;
+        }
+
+        private IntPtr[] GetOrtValuesHandles(IReadOnlyCollection<FixedBufferOnnxValue> values, bool input)
+        {
+            var valuesArray = new IntPtr[values.Count];
+            for (int index = 0; index < values.Count; ++index)
+            {
+                var v = values.ElementAt(index);
+                if (!input && v.ElementType == Tensors.TensorElementType.String)
+                {
+                    throw new NotSupportedException("Using string type FixedBufferOnnxValue in outputs is not supported.");
+                }
+                valuesArray[index] = v.Value.Handle;
+            }
+            return valuesArray;
+        }
+
+        /// <summary>
+        /// Other classes access
+        /// </summary>
+        internal IntPtr Handle
+        {
+            get
+            {
+                return _nativeHandle;
+            }
+        }
+
+    #endregion
+
+    #region IDisposable
+
+        /// <summary>
+        /// Finalizer.
+        /// </summary>
+        ~TrainingSession()
+        {
+            DisposeHelper(false);
+        }
+
+        /// <summary>
+        /// IDisposable implementation
+        /// </summary>
+        public void Dispose()
+        {
+            DisposeHelper(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// IDisposable implementation
+        /// </summary>
+        /// <param name="disposing">true if invoked from Dispose() method</param>
+        protected virtual void DisposeHelper(bool disposing)
```

Comment:
>virtual [](http://example.com/codeflow?start=18&length=7)

I do no think this should be virtual. We do not design this to override it in any of the deriving classes. I think it should be private